iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp

```diff
@@ -0,0 +1,345 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===- ConvertToDestinationPassingStylePass.cpp ---------------------------===//
+//
+// Transformations that are performed before calling upstream Comprehensive
+// Bufferization pass. These change the dispatch region to use destination
+// passing style, mostly to get rid of `init_tensor` ops that result in an
+// allocation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Codegen/Common/BufferizationAnalysis.h"
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Codegen/Transforms/Transforms.h"
+#include "iree/compiler/Codegen/Utils/Utils.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowTypes.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.h"
+#include "iree/compiler/Dialect/Shape/IR/ShapeOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/EquivalenceClasses.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"
+#include "mlir/Dialect/MemRef/Transforms/Passes.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+class ConvertToDestinationPassingStylePass
+    : public ConvertToDestinationPassingStyleBase<
+          ConvertToDestinationPassingStylePass> {
+ public:
+  ConvertToDestinationPassingStylePass() = default;
+  void runOnOperation() override;
+};
+}  // namespace
+
+/// Returns the subview into the buffer that is supposed to be populated with
+/// the `value` of the `flow.dispatch.tensor.store` operation. This can be used
+/// to compute the results in place.
+static Value getTensorLoadOpForTensorStoreOp(OpBuilder &b, Operation *storeOp) {
+  SmallVector<Value, 4> operandsOfLoadOp;
+  auto op = cast<OffsetSizeAndStrideOpInterface>(storeOp);
+  Value target, source;
+  std::tie(source, target) =
+      TypeSwitch<Operation *, std::tuple<Value, Value>>(op)
+          .Case<IREE::Flow::DispatchTensorStoreOp>([&](auto storeOp) {
+            return std::make_tuple(storeOp.value(), storeOp.target());
+          })
+          .Case<tensor::InsertSliceOp>([&](auto storeOp) {
```

Comment:
Functionality for removing `InsertSliceOp`-anchored `InitTensorOp`s has landed in MLIR ComprehensiveBufferize. We can use that directly now. I'll prepare a commit that enables it and we just need to add a unit test.

Reply:
Such `InitTensorOp`s are replaced with `ExtractSliceOp`s. So it's a bit different from what this is doing. The resulting IR should be the same though.

Reply:
Yeah it should end up with the same